### PR TITLE
- Add User entity;

### PR DESCRIPTION
**Pull request by SviatoslavTataryn**
- Add the required DTOs and mapper;
- Add AuthenticationController;
- Add FieldMatch annotation;
- Add GlobalExceptionHandler.

### CODE REVIEW

File: src/main/java/mate/academy/bookstore/model/User.java [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+@Entity
+@Table(name = "Users")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/User.java]:
```suggestion
@Table(name = "users")
```

File: src/main/java/mate/academy/bookstore/model/User.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+public class User {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false, unique = true)
+    private String email;
+    @Column(nullable = false, unique = true)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/User.java]:
```suggestion
    @Column(nullable = false)
```

File: src/main/java/mate/academy/bookstore/service/impl/UserServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+@Service
+@AllArgsConstructor
+public class UserServiceImpl implements UserService {
+    private final AuthenticationService authenticationService;
+    private final UserRepository userRepository;
+    private final UserMapper userMapper;
+
+    @Override
+    public UserResponseDto register(UserRegistrationRequestDto requestDto) {
+        if (authenticationService.authenticate(requestDto)) {
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/service/impl/UserServiceImpl.java]:
```suggestion
        if (userRepository.existsByEmail(requestDto.email())) {
            throw new RegistrationException(
                    String.format("User with this email: %s already exists", requestDto.email())
            );
        }
```

File: src/main/java/mate/academy/bookstore/security/AuthenticationService.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+import org.springframework.stereotype.Service;
+
+@Service
+@AllArgsConstructor
+public class AuthenticationService {
+    private final UserRepository userRepository;
+
+    public boolean authenticate(UserRegistrationRequestDto requestDto) {
+        Optional<User> user = userRepository.findUserByEmail(requestDto.email());
+        return (user.isPresent() && user.get().getPassword().equals(requestDto.password()));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/security/AuthenticationService.java]: You actually don't need this logic (see the UserServiceImpl)

File: src/main/java/mate/academy/bookstore/validation/FieldMatch.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+import jakarta.validation.Payload;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target({ElementType.TYPE, ElementType.ANNOTATION_TYPE})
+@Retention(RetentionPolicy.RUNTIME)
+@Constraint(validatedBy = PasswordValidator.class)
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/validation/FieldMatch.java]:
```suggestion
```

File: src/main/java/mate/academy/bookstore/validation/PasswordValidator.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        firstFieldName = constraintAnnotation.first();
+        secondFieldName = constraintAnnotation.second();
+    }
+
+    @Override
+    public boolean isValid(final Object value,
+                           final ConstraintValidatorContext constraintValidatorContext) {
+        final Object firstObj = new BeanWrapperImpl(value).getPropertyValue(firstFieldName);
+        final Object secondObj = new BeanWrapperImpl(value).getPropertyValue(secondFieldName);
+
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/validation/PasswordValidator.java]:
```suggestion
        return Objects.equals(firstObj, secondObj);
```

File: src/main/resources/db/changelog/changes/02-create-users-table.yaml [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+                  constraints:
+                    unique: true
+                    nullable: false
+              - column:
+                  name: password
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
+              - column:
+                  name: firstName
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/02-create-users-table.yaml]:
```suggestion
                  name: first_name
```

File: src/main/resources/db/changelog/changes/02-create-users-table.yaml [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
+              - column:
+                  name: lastName
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/02-create-users-table.yaml]:
```suggestion
                  name: last_name
```

File: src/main/resources/db/changelog/changes/02-create-users-table.yaml [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+                  type: varchar(255)
+                  constraints:
+                    nullable: false
+              - column:
+                  name: shippingAddress
```
Comment (Elena-Bruyako) [inline src/main/resources/db/changelog/changes/02-create-users-table.yaml]:
```suggestion
                  name: shipping_address
```

Read about naming convention

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Well done! Let's fix some comments

File: src/main/java/mate/academy/bookstore/model/User.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+@Getter
+@Setter
+@Entity
+@Table(name = "users")
+public class User {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/model/User.java]:
```suggestion
    @Column(nullable = false, unique = true)
```

File: src/main/java/mate/academy/bookstore/validation/PasswordValidator.java [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+    @Override
+    public void initialize(FieldMatch constraintAnnotation) {
+        String firstObj = constraintAnnotation.first();
+        String secondObj = constraintAnnotation.second();
+    }
+
+    @Override
+    public boolean isValid(final Object firstObj,
+                           final ConstraintValidatorContext secondObj) {
+        return Objects.equals(firstObj, secondObj);
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstore/validation/PasswordValidator.java:22]: Previous implementation was almost correct (revert)
```suggestion
        final Object firstObj = new BeanWrapperImpl(value).getPropertyValue(firstFieldName);
        final Object secondObj = new BeanWrapperImpl(value).getPropertyValue(secondFieldName);
        return Objects.equals(firstObj, secondObj);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
left some comments

**Review by sarakhmen — APPROVED** [review]

**Pull request merged**